File: SenLib/Sen3/FileFixes/book07_dat.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	class book07_dat : FileMod {
		public string GetDescription() {
			return "Fix typos in Black Records.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x4531fd2f226f5c64ul, 0x1e09fc35e118d1c1ul, 0xbb2a4144u));
			if (file == null) {
				return null;
			}

			var bin = file.CopyToMemory();

			// great knight -> Great Knight
			bin.Position = 0x40a1;
			bin.WriteUInt8(0x47);
			bin.Position = 0x40a7;
			bin.WriteUInt8(0x4b);

			// TODO: Port the CS2 fixes here

			return new FileModResult[] { new FileModResult("data/scripts/book/dat_en/book07.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x4531fd2f226f5c64ul, 0x1e09fc35e118d1c1ul, 0xbb2a4144u));
			if (file == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/book/dat_en/book07.dat", file) };
		}
	}
}

File: native/sen3/file_fixes/book07_dat.cpp
#include <string_view>
#include <vector>

#include "sen/file_getter.h"
#include "p3a/pack.h"
#include "p3a/structs.h"
#include "util/hash/sha1.h"

namespace SenLib::Sen3::FileFixes::book07_dat {
std::string_view GetDescription() {
    return "Fix typos in Black Records.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/book/dat_en/book07.dat",
            34151,
            HyoutaUtils::Hash::SHA1FromHexString("4531fd2f226f5c641e09fc35e118d1c1bb2a4144"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        // great knight -> Great Knight
        bin[0x40a1] = 0x47;
        bin[0x40a7] = 0x4b;

        result.emplace_back(std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::book07_dat
